vpr/src/place/placer.cpp

```diff
@@ -0,0 +1,435 @@
+       }
+
+       log_printer_.print_place_status(temperature_timer.elapsed_sec());
+    }
+    post_quench_timing_stats_ = timing_ctx.stats;
+
+    // Final timing analysis
+    const t_annealing_state& annealing_state = annealer_->get_annealing_state();
+    PlaceCritParams crit_params;
+    crit_params.crit_exponent = annealing_state.crit_exponent;
+    crit_params.crit_limit = placer_opts_.place_crit_limit;
+
+    if (placer_opts_.place_algorithm.is_timing_driven()) {
+       perform_full_timing_update(crit_params, place_delay_model_.get(), placer_criticalities_.get(),
+                                  placer_setup_slacks_.get(), pin_timing_invalidator_.get(),
+                                  timing_info_.get(), &costs_, placer_state_);
+
+       critical_path_ = timing_info_->least_slack_critical_path();
+
+       VTR_LOG("post-quench CPD = %g (ns) \n",
+               1e9 * critical_path_.delay());
+    }
+
+    // See if our latest checkpoint is better than the current placement solution
+    if (placer_opts_.place_checkpointing) {
+       restore_best_placement(placer_state_,
+                              placement_checkpoint_, timing_info_, costs_,
+                              placer_criticalities_, placer_setup_slacks_, place_delay_model_,
+                              pin_timing_invalidator_, crit_params, noc_cost_handler_);
+    }
+
+    if (placer_opts_.placement_saves_per_temperature >= 1) {
+       std::string filename = vtr::string_fmt("placement_%03d_%03d.place",
+                                              annealing_state.num_temps + 1, 0);
+       VTR_LOG("Saving final placement to file: %s\n", filename.c_str());
+       print_place(nullptr, nullptr, filename.c_str(), placer_state_.mutable_block_locs());
+    }
+
+    //#ifdef VERBOSE
+    //    if (getEchoEnabled() && isEchoFileEnabled(E_ECHO_END_CLB_PLACEMENT)) {
+    //        print_clb_placement(getEchoFileName(E_ECHO_END_CLB_PLACEMENT));
+    //    }
+    //#endif
+
+    // Update physical pin values
+    for (const ClusterBlockId block_id : cluster_ctx.clb_nlist.blocks()) {
+       placer_state_.mutable_blk_loc_registry().place_sync_external_block_connections(block_id);
+    }
+
+    check_place_();
+
+    log_printer_.print_post_placement_stats();
+}
+
+void Placer::copy_locs_to_global_state() {
+    auto& place_ctx = g_vpr_ctx.mutable_placement();
+
+    // the placement location variables should be unlocked before being accessed
+    place_ctx.unlock_loc_vars();
+
+    // copy the local location variables into the global state
+    auto& global_blk_loc_registry = place_ctx.mutable_blk_loc_registry();
+    global_blk_loc_registry = placer_state_.blk_loc_registry();
+
+#ifndef NO_GRAPHICS
+    // update the graphics' reference to placement location variables
+    get_draw_state_vars()->set_graphics_blk_loc_registry_ref(global_blk_loc_registry);
+#endif
+}
+
+const PlacementAnnealer& Placer::annealer() const {
```

Comment:
For simple getter methods like these, I recommend just declaring them inline in the header file. Or even consider making them public variables (since most of these are const anyways we do not have to worry about people screwing with variables they are not supposed to).